.github/workflows/generate_lkg.py

```diff
@@ -0,0 +1,216 @@
+
+    Parameters
+    ----------
+    requirements_directory : str
+        Directory containing requirements files
+
+    regex : str
+        Regex to match requirements file names, must have named groups "os" and "pyversion"
+    """
+    req_regex = r'^(?P<pkg>.*?)==(?P<version>.*)$'  # parses requirements from pip freeze results
+    files = listdir(requirements_directory)
+
+    all_combos = set()
+
+    # We'll store the requirements for each version of each package in a dictionary
+    # (e.g. "numpy" -> {0.24.1 -> {Combo1, Combo2, ...}, 0.24.2 -> {Combo3, Combo4, ...}, ...})
+    # each entry of the inner dictionary will become a line in the requirements file
+    # (e.g. "numpy==0.24.1; platform_system=='Linux' and python_version=='3.8' or ...")
+    req_dict = defaultdict(lambda: defaultdict(set))  # package -> package_version -> set of Combos
+
+    for file in files:
+        match = re.match(regex, file)
+        if not match:
+            print(f"Skipping {file} because it doesn't match the regex")
+            continue
+        os = platform_map[match.group('os')]
+        py_version = packaging.version.parse(match.group('pyversion'))
+        combo = Combo(os, py_version)
+        all_combos.add(combo)
+
+        # read each line of the file
+        with open(path.join(requirements_directory, file)) as lines:
+            for line in lines:
+                match = re.search(req_regex, line)
+                pkg_version = packaging.version.parse(match.group('version'))
+                req_dict[match.group('pkg')][pkg_version].add(combo)
+
+    constraint_map, os_map = simple_constraint_map(frozenset(all_combos))
+    # list of all requirements, sorted by package name and version
+    reqs = []
+    for pkg, versions in sorted(req_dict.items()):
+        for version, combos in sorted(versions.items()):
+            combos = frozenset(combos)
+            req = f"{pkg}=={version}"
+
+            if combos in constraint_map:
+                suffix = constraint_map[combos]
+
+            else:
+                # we don't have a simple constraint for this package, so we need to generate a more complex one
+                # which will generally be of the form:
+                # "(platform_system=='os1' and (python_version=='py1' or python_version=='py2') or ...) or
+                #  (platform_system=='os2' and (python_version=='py3' or ...) ..."
+                #
+                # that is, we will OR together constraints grouped by os
+                # for some oses, we might find a nice representation for their python version constraints in the os_map
+                # (e.g. "python_version=='3.8'", or "python_version<'3.8'"), in which case we'll use that;
+                # for others, we'll have to OR together all of the relevant individual versions
+                os_constraints = []
+
+                os_versions = defaultdict(set)  # dictionary from os to set of python versions
+                for fp in combos:
```

Comment:
What do you think about changing occurrences of `fp` into `combo`. Since fp refers to `FileParts` which is now called `Combo` 

For this line specifically it could be 
```
for combo in combos:
```

Reply:
Or maybe even `c` if `combo` is too verbose. Mostly I just think `fp` is vestigial

Reply:
Fixed, good catch.